src/descriptor/tr.rs

```diff
@@ -490,79 +490,114 @@ where
+                if item.n_children_yielded == 0 {
+                    match item.node.parens {
+                        Parens::Curly => {
+                            if !item.node.name.is_empty() {
+                                return Err(Error::Parse(ParseError::Tree(
+                                    ParseTreeError::IncorrectName {
+                                        actual: item.node.name.to_owned(),
+                                        expected: "",
+                                    },
+                                )));
+                            }
+                            if round_paren_depth > 0 {
+                                return Err(Error::Parse(ParseError::Tree(
+                                    ParseTreeError::IllegalCurlyBrace {
+                                        pos: item.node.children_pos,
+                                    },
+                                )));
+                            }
+                        }
+                        Parens::Round => round_paren_depth += 1,
+                        _ => {}
                     }
-                    Tr::new(expression::terminal(key, Pk::from_str)?, None)
                 }
-                2 => {
-                    let key = &top.args[0];
-                    if !key.args.is_empty() {
-                        return Err(Error::Unexpected(format!(
-                            "#{} script associated with `key-path` while parsing taproot descriptor",
-                            key.args.len()
-                        )));
+                if item.is_complete {
+                    if item.node.parens == Parens::Curly {
+                        if item.n_children_yielded == 2 {
+                            let rchild = tree_stack.pop().unwrap();
+                            let lchild = tree_stack.pop().unwrap();
+                            tree_stack.push(TapTree::combine(lchild, rchild));
+                        } else {
+                            return Err(Error::Parse(ParseError::Tree(
+                                ParseTreeError::IncorrectNumberOfChildren {
+                                    description: "Taptree node",
+                                    n_children: item.n_children_yielded,
+                                    minimum: Some(2),
+                                    maximum: Some(2),
+                                },
+                            )));
+                        }
+                    } else {
+                        if item.node.parens == Parens::Round {
+                            round_paren_depth -= 1;
+                        }
+                        if round_paren_depth == 0 {
+                            let script = Miniscript::from_tree(item.node)?;
+                            // FIXME hack for https://github.com/rust-bitcoin/rust-miniscript/issues/734
+                            if script.ty.corr.base != crate::miniscript::types::Base::B {
```

Comment:
We could also yet another tagged type to Miniscript<Pk, State> where State = {Constructing, Finalized}

Reply:
I think that, once I finally get to the end of this mega-PR series and we have a clean validation-parameter structure, I'll be able to remove hacks like this and the user will see a coherent/simple model. Hopefully I don't need an extra generic.